[RPC] Add generatetoaddress rpc to mine to an address

achow101 opened the pull request:
Add a new optional parameter to the generate call to mine to a specific address even if it doesn't exist in the wallet.

achow101 committed:
Create generatetoaddress rpc

Creates the generatetoaddress rpc which is virtually identical to the generate rpc except that it takes an argument for the address to mine to. It does not rely on wallet functionality.

The mining code shared by generate and generatetoaddress has been moved to another method to reduce duplication.

jonasschnelli commented on src/rpc/mining.cpp:
```diff
@@ -114,13 +115,14 @@ UniValue getgenerate(const UniValue& params, bool fHelp)
             "\nNote: this function can only be used on the regtest network\n"
             "\nArguments:\n"
             "1. numblocks    (numeric, required) How many blocks are generated immediately.\n"
+            "2. address    (string, optional) The base58 encoded address to send the newly generated bitcoin to.\n"
```
Maybe use "address" instead of "base58 encoded address" (it would even be wrong because it's a base58 check encoded ripemd hash".

jonasschnelli commented:
Nice!
utACK.

sipa commented:
I like the concept, but perhaps we can separate that out to a different RPC (which then can function without wallet support)?

Also, the change in arguments to `generate` conflicts with #7663.

achow101 commented:
@sipa How about moving it to a new rpc `generatetoaddress`?

luke-jr commented:
This seems unnecessary. Why not just move GBT->block code from getblocktemplate_proposals.py to blocktools?

achow101 commented:
>This seems unnecessary.

Why?

>Why not just move GBT->block code from getblocktemplate_proposals.py to blocktools?

Not sure what you mean by this.

laanwj commented:
Concept ACK
Needs rebase after #7507,#7663

achow101 commented:
Rebased

MarcoFalke commented:
I'd prefer to have rpc tests as part of this pull.

achow101 commented:
@MarcoFalke Where would be the best place to put such tests?

MarcoFalke commented:
@achow101 You can either start from scratch (c.f. https://github.com/bitcoin/bitcoin/blob/48f39058315c908c360acb596923cbc090119480/qa/rpc-tests/disablewallet.py) or extend an existing python script.

sipa commented:
One of the primary use cases for this, I think, would be to have RPC tests where you get an address from one node, and let another node mine directly to it, without the need to generate 100 blocks in between before you can transfer.

achow101 committed:
RPC tests for generatetoaddress

Adds two RPC tests for the generatetoaddress RPC, one in the wallet, and one when the wallet is disabled.

The wallet RPC Test mines Bitcoin to another node's address and checks that that node has received the Bitcoin.

The RPC test without the wallet mines Bitcoin to an arbitrary address and checks that it works. It then mines to an arbitrary invalid address and checks that that fails.

achow101 commented:
I added rpc tests, both with and without wallets.

> The wallet RPC Test mines Bitcoin to another node's address and checks that that node has received the Bitcoin.

> The RPC test without the wallet mines Bitcoin to an arbitrary address and checks that it works. It then mines to an arbitrary invalid address and checks that that fails.

MarcoFalke commented on qa/rpc-tests/wallet.py:
```diff
@@ -249,6 +249,19 @@ def run_test (self):
+            self.nodes[1].gettransaction(cbTxId)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Invalid or non-wallet transaction id" in errorString, False)
```
Nit:

```python
assert("Invalid or non-wallet transaction id" not in errorString)

MarcoFalke replied:
Actually, you can remove the whole `except` block?

achow101 replied:
> Actually, you can remove the whole except block?

Why?

MarcoFalke replied:
The error string is always empty, so it's dead code? (Correct me if I am wrong)

achow101 replied:
Uh, I don't think so. There is an error from gettransaction and the error string was defined earlier.

Anyways, this should work, right?
```
except JSONRPCException,e:
            assert("Invalid or non-wallet transaction id" not in e.error['message'])
```

MarcoFalke commented on qa/rpc-tests/disablewallet.py:
```diff
@@ -29,5 +29,22 @@ def run_test (self):
+            self.nodes[0].generatetoaddress(1, '3J98t1WpEZ73CNmQviecrnyiWrnqRhWNLy')
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Invalid address" in errorString, True)
```
Nit: Move it up one line and make it

```python
            raise AssertionError("Must not mine to invalid address!")
        except JSONRPCException,e:
            assert("Invalid address" in e.error['message'])

achow101 commented:
addressed nits

laanwj commented:
Random Q: Should this also be able to take an arbitrary scriptPubKey?

> One of the primary use cases for this, I think, would be to have RPC tests where you get an address from one node, and let another node mine directly to it,

Also eventually we want to have at least part of the RPC tests usable without wallet.

achow101 commented:
> Random Q: Should this also be able to take an arbitrary scriptPubKey?

Do you think that it should?

laanwj commented:
> Do you think that it should?

I don't deeply care myself. Just a question that came up with me, usually the discussion of custom scripts comes up immediately after allowing arbitrary addresses, as addresses map to a only a subset of the possible output scripts.

achow101 commented:
OK, I'll just leave it as is.

sipa commented:
If we want to support that, perhaps we should just define an address format
that supports arbitrary scriptPubKeys. That way, it would be available
everywhere.

laanwj commented:
Needs rebase.

achow101 commented:
rebased

MarcoFalke commented on qa/rpc-tests/disablewallet.py:
```diff
@@ -29,5 +29,20 @@ def run_test (self):
+            assert("ProcessNewBlock, block not accepted" not in e.error['message'])
+            assert("Couldn't create new block" not in e.error['message'])
+
+        errorString = ''
```
Please remove this line. The variable is unused and I'd prefer not to use it ever again: see 3b4324b

achow101 replied:
fixed

laanwj commented:
ACK fe00ca7
